e2e usecase test failed: HTTPRouteRequestHeaderControl

**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via [ASRC](https://security.alibaba.com/)(Alibaba Security Response Center) where the issue will be triaged appropriately.**

- [x] I have searched the [issues](https://github.com/alibaba/higress/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description

1. [HTTPRouteRequestHeaderControl](https://github.com/alibaba/higress/blob/main/test/ingress/conformance/tests/httproute-request-header-control.go#LL28C16-L28C16) 测试用例未被添加到 [e2e_test.go/higressTests Slice](https://github.com/alibaba/higress/blob/main/test/ingress/e2e_test.go#L53)；

2. 当我尝试添加该用例并执行 `make ingress-conformance-test` 进行测试时，发现该用例出错（其他用例均通过）。

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
$ make ingress-conformance-test
<Snip ...>

Sending Request:
< GET /foo1 HTTP/1.1
< Host: foo.com
< User-Agent: Go-http-client/1.1
< X-Echo-Set-Header:
< Accept-Encoding: gzip
<
<

Received Response:
< HTTP/1.1 200 OK
< Content-Length: 939
< Content-Type: application/json
< Date: Sat, 20 May 2023 07:37:21 GMT
< Req-Arrive-Time: 1684568241296
< Req-Cost-Time: 1
< Resp-Start-Time: 1684568241297
< Server: istio-envoy
< X-Content-Type-Options: nosniff
< X-Envoy-Upstream-Service-Time: 1
<
< {
<  "path": "/foo1",
<  "host": "foo.com",
<  "method": "GET",
<  "proto": "HTTP/1.1",
<  "headers": {
<   "Accept-Encoding": [
<    "gzip"
<   ],
<   "Original-Host": [
<    "foo.com"
<   ],
<   "Req-Start-Time": [
<    "1684568241296"
<   ],
<   "Stage": [
<    "test"
<   ],
<   "User-Agent": [
<    "Go-http-client/1.1"
<   ],
<   "X-B3-Sampled": [
<    "0"
<   ],
<   "X-B3-Spanid": [
<    "d96432e6d0cb75b3"
<   ],
<   "X-B3-Traceid": [
<    "a2ac9323999cd7f1d96432e6d0cb75b3"
<   ],
<   "X-Echo-Set-Header": [
<    ""
<   ],
<   "X-Envoy-Attempt-Count": [
<    "1"
<   ],
<   "X-Envoy-Decorator-Operation": [
<    "infra-backend-v1.higress-conformance-infra.svc.cluster.local:8080/foo1"
<   ],
<   "X-Envoy-Internal": [
<    "true"
<   ],
<   "X-Forwarded-For": [
<    "172.19.0.1"
<   ],
<   "X-Forwarded-Proto": [
<    "http"
<   ],
<   "X-Request-Id": [
<    "a6072530-a0e0-4ca1-a38e-ee1c58a0cb83"
<   ]
<  },
<  "namespace": "higress-conformance-infra",
<  "ingress": "",
<  "service": "",
<  "pod": "infra-backend-v1-55b649759b-pz65r"
< }

    http.go:190: Response expectation failed for request: {{http   localhost /foo1  false false   } foo.com HTTP GET map[X-Echo-Set-Header:[]] false}  not ready yet: expected path to be , got /foo1 (after 29.134586641s)
    http.go:171: timeout while waiting after 30 attempts, 0/3 sucessess
=== NAME  TestHigressConformanceTests/HTTPRouteRequestHeaderControl
    apply.go:158: Deleting httproute-request-header-control-remove Ingress
    apply.go:158: Deleting httproute-request-header-control-update Ingress
    apply.go:158: Deleting httproute-request-header-control-add-more Ingress
    apply.go:158: Deleting httproute-request-header-control-add-one Ingress
=== NAME  TestHigressConformanceTests
    apply.go:114: Deleting tls-validity-checks-certificate
    apply.go:114: Deleting certificate
    apply.go:158: Deleting web-backend Deployment
    apply.go:158: Deleting web-backend Service
    apply.go:158: Deleting higress-conformance-web-backend Namespace
    apply.go:158: Deleting app-backend-v2 Deployment
    apply.go:158: Deleting app-backend-v2 Service
    apply.go:158: Deleting app-backend-v1 Deployment
    apply.go:158: Deleting app-backend-v1 Service
    apply.go:158: Deleting higress-conformance-app-backend Namespace
    apply.go:158: Deleting infra-backend-v3 Deployment
    apply.go:158: Deleting infra-backend-v3 Service
    apply.go:158: Deleting infra-backend-v2 Deployment
    apply.go:158: Deleting infra-backend-v2 Service
    apply.go:158: Deleting infra-backend-v1 Deployment
    apply.go:158: Deleting infra-backend-v1 Service
    apply.go:158: Deleting higress-conformance-infra Namespace
--- FAIL: TestHigressConformanceTests (91.04s)
    --- FAIL: TestHigressConformanceTests/HTTPRouteRequestHeaderControl (30.71s)
        --- FAIL: TestHigressConformanceTests/HTTPRouteRequestHeaderControl/Request_header_control (30.00s)
FAIL
FAIL        command-line-arguments        91.062s
FAIL
make: *** [Makefile.core.mk:223: run-ingress-e2e-test] Error 1

```


### Ⅲ. Describe what you expected to happen


### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅴ. Anything else we need to know?


### Ⅵ. Environment:

- Higress version:  
- OS :
- Others:

--- a/test/ingress/conformance/tests/httproute-request-header-control.go
+++ b/test/ingress/conformance/tests/httproute-request-header-control.go
@@ -32,6 +32,12 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
 		testcases := []http.Assertion{
 			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 1: add one",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
 				Request: http.AssertionRequest{
 					ActualRequest: http.Request{
 						Path: "/foo1",
@@ -39,6 +45,8 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 					},
 					ExpectedRequest: &http.ExpectedRequest{
 						Request: http.Request{
+							Path: "/foo1",
+							Host: "foo.com",
 							Headers: map[string]string{
 								"stage": "test",
 							},
@@ -53,6 +61,12 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 				},
 			},
 			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 2: add more",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
 				Request: http.AssertionRequest{
 					ActualRequest: http.Request{
 						Path: "/foo2",
@@ -60,6 +74,8 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 					},
 					ExpectedRequest: &http.ExpectedRequest{
 						Request: http.Request{
+							Path: "/foo2",
+							Host: "foo.com",
 							Headers: map[string]string{
 								"stage":  "test",
 								"canary": "true",
@@ -75,6 +91,12 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 				},
 			},
 			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 3: update one",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
 				Request: http.AssertionRequest{
 					ActualRequest: http.Request{
 						Path: "/foo3",
@@ -85,9 +107,10 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 					},
 					ExpectedRequest: &http.ExpectedRequest{
 						Request: http.Request{
+							Path: "/foo3",
+							Host: "foo.com",
 							Headers: map[string]string{
-								"stage":  "pro",
-								"canary": "true",
+								"stage": "pro",
 							},
 						},
 					},
@@ -100,16 +123,91 @@ var HTTPRouteRequestHeaderControl = suite.ConformanceTest{
 				},
 			},
 			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 4: update more",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
 				Request: http.AssertionRequest{
 					ActualRequest: http.Request{
 						Path: "/foo4",
+						Host: "foo.com",
+						Headers: map[string]string{
+							"stage":  "test",
+							"canary": "true",
+						},
+					},
+					ExpectedRequest: &http.ExpectedRequest{
+						Request: http.Request{
+							Path: "/foo4",
+							Host: "foo.com",
+							Headers: map[string]string{
+								"stage":  "pro",
+								"canary": "false",
+							},
+						},
+					},
+				},
+
+				Response: http.AssertionResponse{
+					ExpectedResponse: http.Response{
+						StatusCode: 200,
+					},
+				},
+			},
+			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 5: remove one",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
+				Request: http.AssertionRequest{
+					ActualRequest: http.Request{
+						Path: "/foo5",
 						Host: "foo.com",
 						Headers: map[string]string{
 							"stage": "test",
 						},
 					},
 					ExpectedRequest: &http.ExpectedRequest{
+						Request: http.Request{
+							Path: "/foo5",
+							Host: "foo.com",
+						},
 						AbsentHeaders: []string{"stage"},
+					},
+				},
+
+				Response: http.AssertionResponse{
+					ExpectedResponse: http.Response{
+						StatusCode: 200,
+					},
+				},
+			},
+			{
+				Meta: http.AssertionMeta{
+					TestCaseName:    "case 6: remove more",
+					TargetBackend:   "infra-backend-v1",
+					TargetNamespace: "higress-conformance-infra",
+				},
+
+				Request: http.AssertionRequest{
+					ActualRequest: http.Request{
+						Path: "/foo6",
+						Host: "foo.com",
+						Headers: map[string]string{
+							"stage":  "test",
+							"canary": "true",
+						},
+					},
+					ExpectedRequest: &http.ExpectedRequest{
+						Request: http.Request{
+							Path: "/foo6",
+							Host: "foo.com",
+						},
+						AbsentHeaders: []string{"stage", "canary"},
 					},
 				},
 

--- a/test/ingress/conformance/tests/httproute-request-header-control.yaml
+++ b/test/ingress/conformance/tests/httproute-request-header-control.yaml
@@ -61,7 +61,7 @@ kind: Ingress
 metadata:
   annotations:
     higress.io/request-header-control-update: stage pro
-  name: httproute-request-header-control-update
+  name: httproute-request-header-control-update-one
   namespace: higress-conformance-infra
 spec:
   ingressClassName: higress
@@ -81,8 +81,10 @@ apiVersion: networking.k8s.io/v1
 kind: Ingress
 metadata:
   annotations:
-    higress.io/request-header-control-remove: stage
-  name: httproute-request-header-control-remove
+    higress.io/request-header-control-update: |
+      stage pro
+      canary false
+  name: httproute-request-header-control-update-more
   namespace: higress-conformance-infra
 spec:
   ingressClassName: higress
@@ -92,6 +94,48 @@ spec:
         paths:
           - pathType: Exact
             path: "/foo4"
+            backend:
+              service:
+                name: infra-backend-v1
+                port:
+                  number: 8080
+---
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  annotations:
+    higress.io/request-header-control-remove: stage
+  name: httproute-request-header-control-remove-one
+  namespace: higress-conformance-infra
+spec:
+  ingressClassName: higress
+  rules:
+    - host: "foo.com"
+      http:
+        paths:
+          - pathType: Exact
+            path: "/foo5"
+            backend:
+              service:
+                name: infra-backend-v1
+                port:
+                  number: 8080
+---
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  annotations:
+    higress.io/request-header-control-remove: stage,canary
+  name: httproute-request-header-control-remove-more
+  namespace: higress-conformance-infra
+spec:
+  ingressClassName: higress
+  rules:
+    - host: "foo.com"
+      http:
+        paths:
+          - pathType: Exact
+            path: "/foo6"
             backend:
               service:
                 name: infra-backend-v1

--- a/test/ingress/e2e_test.go
+++ b/test/ingress/e2e_test.go
@@ -72,6 +72,7 @@ func TestHigressConformanceTests(t *testing.T) {
 		tests.HTTPRouteMatchPath,
 		tests.HttpForceRedirectHttps,
 		tests.HttpRedirectAsHttps,
+		tests.HTTPRouteRequestHeaderControl,
 	}
 
 	cSuite.Run(t, higressTests)

